### PR TITLE
Issue 316

### LINKED ISSUE #316: Scan Cookies On Page Load | Create Cookie Label
As discussed in #314, with the pending changes to `Request` objects, we will no longer catch cookies that are set in request/response headers.

However, we can still look at cookies without catching them in the headers via the cookies API.
> My prefered solution for the cookie issue would be to use the [cookies API](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/cookies/getAll) to scan through a page's cookies for watchlist items/coordinates. We could do this scan on the [page loaded event](https://developer.mozilla.org/en-US/docs/Web/API/Window/load_event). I think this is lower cost than adding headers back. Also, this would allow us to add a `cookie` label if we find information there. 

The benefit of this solution is that we can:
- run specifically the watchlist/coordinate routines on cookies
- reduce the complexity of the overall analysis
- Give an additional `cookie` label when information is found in this routine.

So, for the backend, the tasks are:
- Interact with the cookies API to access a website's cookies
- Set up an analysis flow for these cookies
- Set up an event listener to initiate this flow on page load
- Add `cookie: true` property to evidence objects created from this routine

For the frontend, we should make this minimally complex:
- Add `cookie` badge for evidence objects with this property (Maybe this emoji appended to the end of the badge:  🍪 )
- Some kind of explanation either on hover for the cookie emoji or somewhere else.

So maybe a row of badges could look like this: **( Fine Location 🍪 ) (State)**

I think this choice in UI makes the most sense because `cookie` is a trait of the specific Evidence object; not higher up like `reqUrl` or `permission`.

### PR DESCRIPTION
**Pull request by danielgoldelman**
Searching for Cookies containing evidence we would like to display to the user.

Close #316 on merge.

### CODE REVIEW

File: src/background/analysis/analyze.js [hunk 1]
```diff
@@ -144,6 +145,15 @@ function resolveBuffer(id, data) {
 async function analyze(request, userData) {
 
   const allEvidence = getAllEvidenceForRequest(request, userData);
+
+  if (await browser.cookies.getAll({url: request.reqUrl}).length != 0){
```
Comment (notowen333) [inline src/background/analysis/analyze.js:149]: It looks like you're making the same call, `await browser.cookies.getAll({url: request.reqUrl}`, in this line and then also in the one below. Better to do it once and save the result.

File: src/background/analysis/requestAnalysis/scanCookies.js [hunk 2]
```diff
@@ -0,0 +1,121 @@
+/*
+Licensed per https://github.com/privacy-tech-lab/privacy-pioneer/blob/main/LICENSE
+privacy-tech-lab, https://www.privacytechlab.org/
+*/
+
+import { regexSearch, coordinateSearch, locationKeywordSearch, ipSearch, encodedEmailSearch } from "./searchFunctions.js"
+import { permissionEnum, typeEnum } from "../classModels.js"
+
+/**
+ * This function runs all of the apporpriate analysis functions for an HTTP request.
```
Comment (notowen333) [inline src/background/analysis/requestAnalysis/scanCookies.js:10]: Update comment to describe that this is cookie analysis.

File: src/libs/label-detail/components/item/index.js [hunk 3]
```diff
@@ -57,6 +57,7 @@ const Item = ({ request, url, label }) => {
         {Object.entries(request).map(([type, request]) => (
           <SBadge key={type} className="badge" onClick={(event) => inflateCollapse(event, request, type)}>
             {privacyLabels[label]["types"][type]["displayName"]}
+            {request.cookie?` 🍪`:null}
```
Comment (notowen333) [inline src/libs/label-detail/components/item/index.js:60]: Nice! I love how concise this UI is.

File: src/libs/label-detail/components/evidence/index.js [hunk 4]
```diff
@@ -41,17 +41,25 @@ const Evidence = ({ collapseId, request, label, type }) => {
...
+      } else {
+        const maxChars = 475
+        const data = { leading: "", middle: "", trailing: "" }
+        data.middle = request.snippet.slice(request.index[0], request.index[1])
+
+        data.leading = request.snippet.slice(0, request.index[0])
+        data.trailing = request.snippet.slice(request.index[1], request.snippet.length)
+
+        data.leading = "... " + data.leading.slice(maxChars * -1)
+        data.trailing = data.trailing.slice(0, maxChars) + " ..."
```
Comment (notowen333) [inline src/libs/label-detail/components/evidence/index.js:59]: Unrelated to your PR but is this necessary? We already cut down the snippet in `cutDownSnippet` in `addEvidence.js`.

File: src/background/analysis/analyze.js [hunk 5]
```diff
@@ -144,6 +145,15 @@ function resolveBuffer(id, data) {
 async function analyze(request, userData) {
 
   const allEvidence = getAllEvidenceForRequest(request, userData);
+
```
Comment (notowen333) [inline src/background/analysis/analyze.js:148]: Another thought here is that we will be analyzing the same cookies over and over the way you have it set up. I think we should maintain a map of reqUrls -> timestamps for a given browsing session. Then we should only run the cookie analysis once every minute (or even 20 seconds or something). It's just clear that we will run the same analysis for the main first-party cookies many times in a row on page load.

**Review by notowen333 — CHANGES_REQUESTED** [review]
Great PR! 

So there are the two small changes
- duplicate call in `analyze.js`
- documentation for `scanCookies.js`

and then the one bigger one:
- maintain a map of reqUrls -> timestamps so we don't run the same cookie analysis many times over.

File: src/libs/label-detail/components/evidence/index.js (hunk 4 above)
Comment (danielgoldelman) [reply src/libs/label-detail/components/evidence/index.js:59]: It would matter if we have a user that turns on "save full snippets"

**Review by Lr-Brown — APPROVED** [review]

**Comment by danielgoldelman** [comment]
^ change from using session storage to using an object that is updated periodically.

@notowen333 I implemented this version twice: once with looking only at the host name, once with looking at the entire request url. When only looking at the host name as the key, the analysis was run once a minute as desired, but missed some cookies. When the entire request url was the key, it run the analysis on the same cookies multiple times, but also caught all of the cookies including keywords we wanted to find. Let's discuss during today's meeting.

**Comment by SebastianZimmeck** [comment]
@danielgoldelman, I am assuming runtime performance is the issue here. How was the performance when the entire request URL was the key?

**Comment by danielgoldelman** [comment]
^ change to using ```domain``` as opposed to ```name``` in ```cookies.getAll```. This grabs the domain name uniquely while still getting all cookies associated. ```name``` only grabbed per request url, thus missing some cookies or repeating analysis.

**Comment by notowen333** [comment]
Looks great! 

@danielgoldelman I made some final changes to neaten the code. In a control flow, if you have the exact same code in two or more places, try to change the flow to avoid this.

I'm very happy with the cookie UI. I think (I hope) users will understand what this means!

Also, here is finally an example where coarse location correctly flags while fine location misses.

Screen Shot 2021-09-12 at 11 05 22 PM

**Review by notowen333 — APPROVED** [review]

**Comment by SebastianZimmeck** [comment]
Could we change latittude to latitude?

If we find lat/lon data, shouldn't that be fine location?

Also, I am not sure what degrees mean. Is that understandable to an ordinary user? Take that out?

(I believe this can be addressed outside of a PR and directly in main as it is only minor work.)

**Comment by SebastianZimmeck** [comment]
Also, Context Below seems redundant to me. Can we take that out?

**Comment by notowen333** [comment]
@SebastianZimmeck 
>Could we change latittude to latitude?

- I will commit this to main now.

>If we find lat/lon data, shouldn't that be fine location?

- These coordinates were labeled as "coarse location" because the lat/lon found were more than .1 degrees  (~10 miles)  from my actual coordinates. We can talk about the best way to describe this difference on Tuesday. 

- It is also an option to remove the coarse location category (which is much much more prone to false positives) and maybe lessen the "fine location" bound slightly.

>Also, I am not sure what degrees mean. Is that understandable to an ordinary user? Take that out?

- Degrees are degrees of latitude/longitude. I agree that this is not clear. This gets back to the above point about sorting out coarse vs fine.

>Also, Context Below seems redundant to me. Can we take that out?

- Let's discuss this Tuesday. Maybe we remove this but also rename "Data Snippet" to "HTTP Context"? I want to keep the word "context" because the idea is to take a look at the context in which we flagged something. 
-  Then, stylistically, I think the emoji add some flare (the down pointer could be moved elsewhere), but I might be alone in this opinion :)

**Comment by SebastianZimmeck** [comment]
OK, great! Thanks, @notowen333. Let's discuss in our next meeting.

**Pull request merged**